File: Java-Core/src/main/java/presto/expression/CastExpression.java
package presto.expression;

import presto.error.PrestoError;
import presto.error.SyntaxError;
import presto.runtime.Context;
import presto.type.IType;
import presto.utils.CodeWriter;
import presto.value.IValue;

public class CastExpression implements IExpression {
	
	IExpression expression;
	IType type;
	
	public CastExpression(IExpression expression, IType type) {
		this.expression = expression;
		this.type = type;
	}
	
	@Override
	public IType check(Context context) throws SyntaxError {
		IType actual = expression.check(context);
		if(!type.isAssignableTo(context, actual))
			throw new SyntaxError("Cannot cast " + actual.toString() + " to " + type.toString());
		return type;
	}

	@Override
	public IValue interpret(Context context) throws PrestoError {
		return expression.interpret(context);
	}

	@Override
	public void toDialect(CodeWriter writer) {
		switch(writer.getDialect()) {
		case E:
		case P:
			expression.toDialect(writer);
			writer.append(" as ");
			type.toDialect(writer);
			break;
		case O:
			writer.append("(");
			type.toDialect(writer);
			writer.append(")");
			expression.toDialect(writer);
			break;
		}
		
	}

}

File: Java-CodeServer/src/main/resources/js/prompto/expression/CastExpression.js
var SyntaxError = require("../error/SyntaxError").SyntaxError;

function CastExpression(expression, type) {
    this.expression = expression;
    this.type = type;
    return this;
}

CastExpression.prototype.check = function(context) {
    var actual = this.expression.check(context);
    if(!this.type.isAssignableTo(context, actual))
        throw new SyntaxError("Cannot cast " + actual.toString() + " to " + type.toString());
    return this.type;
};

CastExpression.prototype.interpret = function(context) {
    return this.expression.interpret(context);
};

CastExpression.prototype.toDialect = function(writer) {
    writer.toDialect(this);
};

CastExpression.prototype.toEDialect = function(writer) {
    this.expression.toDialect(writer);
    writer.append(" as ");
    this.type.toDialect(writer);
};

CastExpression.prototype.toSDialect = function(writer) {
    this.toEDialect(writer);
};

CastExpression.prototype.toODialect = function(writer) {
    writer.append("(");
    this.type.toDialect(writer);
    writer.append(")");
    this.expression.toDialect(writer);
};

exports.CastExpression = CastExpression;
